java/ql/src/Telemetry/ExternalAPI.qll

```diff
@@ -0,0 +1,52 @@
+/** Provides classes and predicates related to handling APIs from external libraries. */
+
+private import java
+private import APIUsage
+private import semmle.code.java.dataflow.ExternalFlow
+
+/**
+ * An external API from either the Java Standard Library or a 3rd party library.
+ */
+class ExternalAPI extends Callable {
+  ExternalAPI() { not this.fromSource() }
+
+  /** Holds true if this API is part of a common testing library or framework */
+  predicate isTestLibrary() { getDeclaringType() instanceof TestLibrary }
+
+  /**
+   * Gets information about the external API in the form expected by the CSV modeling framework.
+   */
+  string asCSV(ExternalAPI api) {
+    result =
+      api.getDeclaringType().getPackage() + ";?;" + api.getDeclaringType().getSourceDeclaration() +
+        ";" + api.getName() + ";" + paramsString(api)
+  }
+
+  /** Holds true if this API is not yet supported by existing CodeQL libraries */
+  predicate isSupported() { not supportKind(this) = "?" }
+
+  /**
+   * Gets the jar file containing this API. Normalizes the Java Runtime to "rt.jar" despite the presence of modules.
+   */
+  string jarContainer() { result = containerAsJar(this.getCompilationUnit().getParentContainer*()) }
```

Comment:
Could you explain what the transitive closure achieves here? It's not immediately clear to me.

Reply:
The immeidate parent of the CUs is usually a `Folder`, either a module like `/modules/java.base/java/util/function` or a package within a jar `/opt/src/community/record-storage-engine/target/neo4j-record-storage-engine-4.3.3-SNAPSHOT.jar/org/neo4j/kernel/impl/store`.